### PR TITLE
[SPARK-21344][SQL] BinaryType comparison does signed byte array comparison

### PR DESCRIPTION
**Pull request by kiszk**
## What changes were proposed in this pull request?

This PR fixes a wrong comparison for `BinaryType`. This PR enables unsigned comparison and unsigned prefix generation for an array for `BinaryType`. Previous implementations uses signed operations.

## How was this patch tested?

Added a test suite in `OrderingSuite`.

### CODE REVIEW

**Comment by srowen** [comment]
Aren't byte comparisons signed though elsewhere? as well as other integer type comparison? this would become inconsistent. Maybe I'm missing the reason this should be different.

**Comment by kiszk** [comment]
IIUC, `BinaryType` should be handled as unsigned byte while Java and Scala have only signed byte array.
That is why we have to do unsigned comparison of `BinaryType`.
The places where byte comparison is used rather than for `BinaryType`, I think that we do not change comparisons.

What do you think?
cc: @gatorsmile

**Comment by maropu** [comment]
I checked other dbms (dbms-like) systems handled binary as unassigned;
```
postgres=# SELECT E'\\x00'::BYTEA < E'\\xff'::BYTEA;
 ?column? 
----------
 t
(1 row)

hive> SELECT unhex('00') < unhex('ff');
OK
true

mysql> SELECT x'00' < x'ff';
+---------------+
| x'00' < x'ff' |
+---------------+
|             1 |
+---------------+
```

In some parts of the Spark implementation, they also handle binary as unassigned (https://github.com/apache/spark/blob/master/core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PrefixComparators.java#L36). I feel, based on the principle of least astonishment, we'd be better to handle binary as unassigned in user-facing pieces.

**Comment by kiszk** [comment]
@maropu, thank you for your comment.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/OrderingSuite.scala [hunk 1]
```diff
@@ -137,4 +137,22 @@ class OrderingSuite extends SparkFunSuite with ExpressionEvalHelper {
     // verify that we can support up to 5000 ordering comparisons, which should be sufficient
     GenerateOrdering.generate(Array.fill(5000)(sortOrder))
   }
+
+  test("SPARK-21344: BinaryType comparison does signed byte array comparison") {
+    val b1 = Array[Byte](1)   // 0x01
+    val b2 = Array[Byte](-1)  // 0xff
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/OrderingSuite.scala]: Hi, @kiszk 
Can we have another test case with more than one byte for code coverage?

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/ByteArray.java [hunk 2]
```diff
@@ -44,7 +44,7 @@ public static long getPrefix(byte[] bytes) {
       final int minLen = Math.min(bytes.length, 8);
       long p = 0;
       for (int i = 0; i < minLen; ++i) {
-        p |= (128L + Platform.getByte(bytes, Platform.BYTE_ARRAY_OFFSET + i))
+        p |= ((long)Platform.getByte(bytes, Platform.BYTE_ARRAY_OFFSET + i) & 0xff)
```
Comment (maropu) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/ByteArray.java]: Nit: maybe we need a single space like `((long) Platform...`.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/OrderingSuite.scala (hunk 1 above)
Comment (kiszk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/OrderingSuite.scala]: Thank you for your comment, done.

Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/OrderingSuite.scala]: Also, it looks better to check the boundary values like `-128` and  `127`.

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/ByteArray.java (hunk 2 above)
Comment (kiszk) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/ByteArray.java]: good catch. done.

**Comment by maropu** [comment]
Jenkins, retest this please.

**Comment by kiszk** [comment]
@gatorsmile could you please review this?

**Comment by gatorsmile** [comment]
```SQL
SELECT x'00' < x'0f'
SELECT x'00' < x'ff'
```

Could you add the above queries to a new `comparator.sql` file?

**Review by gatorsmile — APPROVED** [review]
LGTM

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by kiszk** [comment]
@gatorsmile thank you for your comment, done

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/OrderingSuite.scala [hunk 3]
```diff
@@ -137,4 +137,26 @@ class OrderingSuite extends SparkFunSuite with ExpressionEvalHelper {
...
   }
+
+  test("SPARK-21344: BinaryType comparison does signed byte array comparison") {
+    val data = Seq(
+      (Array[Byte](1), Array[Byte](-1)),
+      (Array[Byte](1, 1, 1, 1, 1), Array[Byte](1, 1, 1, 1, -1)),
+      (Array[Byte](1, 1, 1, 1, 1, 1, 1, 1, 1), Array[Byte](1, 1, 1, 1, 1, 1, 1, 1, -1))
+      )
+    data.foreach { case (b1, b2) =>
+      val rowOrdering = InterpretedOrdering.forSchema(Seq(BinaryType, BinaryType))
```
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/OrderingSuite.scala]: hmmm, don't you just have one binary field for each row? Why you specify two fields to compare?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/OrderingSuite.scala [hunk 4]
```diff
@@ -137,4 +137,26 @@ class OrderingSuite extends SparkFunSuite with ExpressionEvalHelper {
...
+      val genOrdering = GenerateOrdering.generate(
+        BoundReference(0, BinaryType, nullable = true).asc ::
+          BoundReference(1, BinaryType, nullable = true).asc :: Nil)
+      val rowType = StructType(
+        StructField("b1", BinaryType, nullable = true) ::
+          StructField("b2", BinaryType, nullable = true) :: Nil)
```
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/OrderingSuite.scala]: You specify two binary fields as row schema. But actually your input just has one binary field (i.e., Row(b1)).

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/ByteArray.java [hunk 5]
```diff
@@ -44,7 +44,7 @@ public static long getPrefix(byte[] bytes) {
       final int minLen = Math.min(bytes.length, 8);
       long p = 0;
       for (int i = 0; i < minLen; ++i) {
-        p |= (128L + Platform.getByte(bytes, Platform.BYTE_ARRAY_OFFSET + i))
+        p |= ((long) Platform.getByte(bytes, Platform.BYTE_ARRAY_OFFSET + i) & 0xff)
```
Comment (viirya) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/ByteArray.java]: Seems we don't have unit test for this. Shall we add a ByteArraySuite?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/OrderingSuite.scala (hunk 3 above)
Comment (kiszk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/OrderingSuite.scala]: Good catch, updated.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/OrderingSuite.scala (hunk 4 above)
Comment (kiszk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/OrderingSuite.scala]: I misunderstood, updated.

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/ByteArray.java (hunk 5 above)
Comment (kiszk) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/ByteArray.java]: We have [the unit test](https://github.com/kiszk/spark/blob/08776e1931ee21e5200a977efd0aa379ae448033/core/src/test/scala/org/apache/spark/util/collection/unsafe/sort/PrefixComparatorsSuite.scala#L62-L115) for this. I added some cases.

**Comment by viirya** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merging to master/2.2/2.1/2.0

**Pull request closed without merging**